### PR TITLE
Improved ts support for MapForm and ListForm

### PR DESCRIPTION
**Pull request by thiemok**
# Why
formalistics current level of typings are currently lacking generic support for the more complex Items `MapForm` and `ListForm`.
These would greatly profit from a generic typing which can represent the structure of their value.
Without this any access of Items contained within a `MapForm` or `ListForm` will require casting, which greately reduces the typesafety and makes code accessing one of them hard to read.

# What
This adds typings for the value structure of `MapForm` and `ListForm`.
Changes can be tested in this [Playground][1]

# TODO
- [x] Basic support for the value structure of `MapForm`
- [x] Basic support for the value structure of `ListForm`
- [x] Improve support for access of nested structured Items (`getIn`, `updateIn` etc.) if possible
- [x] Improve support for partial `MapForm` if possible

[1]:https://www.typescriptlang.org/play?ssl=86&ssc=1&pln=334&pc=11#code/PQKgUABCELIIYAcBiB7ATgWwgFQKYGcAXfSEYMAYxQDsiJCDDVMIBeCCtXOB+ZdDAAoA3pAjiAlgwz4AXBFHilSgEYoUAG27UkE3BoAm8ztwa79BkRABucDQFdc8gGZ38uCAF8AlABoxyuLU9hgquGjmhsZcPLiRlsI2do7yAAwAdACMXn4BXmA+ANxglDR0apra8Wz0jMwY6QDmuISCAOQVWnA6eoZt3sVUtIQQGBL4+BLUjdXsDET1TS3tYxNTM70G-YNlIwBW+DXzTALphCgAUgDKggMlQ3RcBvYUuAZHdadPL7iCgnAUCi+Dj2NBcaiEYEAa1wAE9vGwAHwKPJcQig6gQAEUCAAaggmWKnmBqW8JVKwwgCDgaEIEjseAWAhqJlifHqIjyUlwMnkikC4k6VU20VMcU2VlsDicEFcGncOTyngK3nSCHsrTaUyk9I0GlhMHGk2m8TawNZZgliSlKQgbWc6jaOTuDxGuAAHgxqAY3vUatTabrGSdMGqNe1qLgAO6m80xS0WSXJGUAbQAuj4XbsIFwMChrL7Nn72AG6QzPqHc-nfh11F0ehZthS6FMpRIDAAlHn5uz+mlljTBxZVgvtVt2dumu7NkYGFD2FRaLt5qV9wPlpmV7uj2uVbpT9IjmtC-ebJtgUCkCAAGXGIawwZIUHIrtqRFvm6w7AtuA-96s3K8hAKYcPG4qJtayZpDkcZivESbSvI2Q+BmWaUs4EhoAsmwfO+d6LM0rSksUxx-os6r4AAFoIP7wZBiEQAATM6OyUtqbYGAACvYVG4YQZGnBR1G0VaSQMVq1AcfQsIILgbSZiRjACaGUzuLSggAKywbEdFibaGkKTOEDjho7YAJK0OEIxzEp+GnKpVmadpCaGAhtoSVJhAyXJhlGT6uAIH+ADiLR8cpDSERZggppkabTq+plENcYV2aG5zXLcxRGWM7pvMG4UsmB4UAdIcjAaBcGiTaMqpDBFU6VVUF2vgKAYB4RBoOsTooUU2USLlBizKM-V5bZn5LK0mTTpeUAQAAwq1CBaO6OCME+ZBGVQGBLR6xb1bwiAcvykilXyeTKCeDZRPt4GufRtpygqPj+AKQQhGEEQijdunVWkWQwedSgYIdX0-uyAicq9gSAWVx1Q4EHXrPEooNRBekym0LVtRAiPTPJuTw69NJoHAsLI99jUMSmbQ0HJwJtIQUYoGadqEJRXByWmiqE+IypQ89gPiIlTCg0VqVCIkMPyCBIlo79BJ1bLd3o-IjE9QTr3C+yotisVkuncBgtKGDh0Q-rPKw7KOuo8r8vId4ngC4TJv8JgJUW3yVsWCjLkJCrTE5I7ZJQ6hSoqqxdBbTt7rJd+i3LYs6U3PF2b+QgIXWRw8e7ackXUNFbTA8gZ707jGyNnFEezrgAUZ3+LLZ+6BEtFF1PC7GECpJXfk1+nLR-nwDfbQnuct-nbd3trjbApk9MYRXaEtpJE4GAAIr3GdD9HzeEK3hcg9PdqrMa5d9BA3c9wFACqCAGLEW8j6G9i37Ee9Fx3bTE6TpppsCgjz4YBErBkT8jRBiG64M3ZwyliiAUcAfa3T9vLT+3UNYXQQT9JqHR8Zh15ncHwZxLjJyysAUhZDyEUMoVQ6hNDaF0PoeQMAHoEDoBGF5WSEAzLSBqPEAAPN0WEyIAA+sBTaYH4dQQREARHhQkYIkh4BZrxHWow5hrDjIQnCK4V480wLxAAPIIGILwgAagAQWvFfAAogAfWwAATU4lYkBYh5bmMsbYhxTjiiSHwFYgAjvYOwAB+eQ-x5DuOsXYxxVjgQqAiRYqJXirFAORJdHx9B5wUEom8UJEB0muJXjwdAeTBBuMSZ4mJqSIAmKKXSGgXZ8D2A0IQIkJQ1G0g0QwNA2iPB8MiZUpxLjxAxDnNQfU-sBnRO8WIUZNAJnnBeDkow+S6zaAyXM8ZsIICUT0CTNA2TYTYCycs+QBSRncDGRMtqEw4DNDKrU0yd96nUBgAQfAdzcDpg2Zc+Z2zgbuiuLgAsnUvLyCBSCqQsIflwCudsjiZy1ndBhXC0YcBAXAvCFC-RzgAASeyaSHPBZi0F0LZm-K2TsglBzKKwkee2RFe5qBZXEO4QgjzHCCEjFGDlMopnJO8PIfpFTpnOIyWy45Sy3iCEWdkt4jL6yCogMKjxorEQZMImYvUbzbn3Isvi8IhLaW3AiXUiQNAdUfOaN8sQScTU1JFckjJRdeHYERIIIusk0BhIASslVSSYmImqdgJV2A2lgB9BQDQNIPDOHsNQCgLyKYWFMY6wNggUBGLKnNPRmxDHGP5YGpV-rBlipKDNaAkCMCkFUe6FhnSpjdN6aI12GAuEW1gcBakbN5BlzTPIdtGBw0dJGI2rRAIPA5rFFW-N+BU2quSRAD0XoDCHCrYO-AwzjKlTyYWmZ4hZXLLyecsS7ZiloFKQ6TACSF1VKRDUs1DSCDNNaQUdpdb1Fjp6ROiAQLCCSrlQYfN5raCdq4BQUE7gFXrLfUwj9DbNHfp0VW+dAanFLs9Lgb0a6xFttKluzZCyTnytWUylFfzRjvM+Q8x9ryqPWrTORylAKIVYrBb+klUKmMTIRaR+s3H-notY6SnFBr9lEo45CryAmqWGppUc4jKyT2Ee2bsuThz6VKaRcyko4h1SEF4ZxDDK7Dhl2BGZYzWHV2cOkO67tlF5CcWBIBAdSqb7PLeChvdsSIBOc4YiMQhFDOWewxAGEsIUDOAdbeoZgh7OOaVd5lMnFGOBbHsF5dVnDicR4JRVDpbER2dywl+QAA5RgbxeXYG8vl0VwJOLqvJSuXAGXMOhfC5F6LaHnFxeK75pVKH9FjAM95+rhWfnPFeLwjs7rvivG9RAf4gJ5AdnNKCG93WUwdai953+YW4TyG2110t1TVs4xrisjsSqOwZKGIQOAqkADScJBDhcO3CTr3mlUnufh5iyrWTO+dy7V5JiJzMhes4OorPbfPAl+7EBbggXMQHK0QSryZquyRBzEsb1SzJKtRwwAw7nYgGC82mpx9XzONdZS0ADyyZWKeg90YEmbiB5L-fTt4wGygDdw9joZGqWhao0Ja6j+rqWHPtZpngIGxcMYyXapVg6qs1fJzF5xjXlQRtwFGmNsp42JpAxA-n3mIc4dbRu91bP8B5KnWyXDs6BfOL56253jWLyKOgOFGt7762jsQ82+3DBwpO7N5l0Lg70xbphruinVjFeKePdpjJHFz2XoEBtk796ZcvMaS+4d8GA9Np-bI8PbXIfSGj2BilEzJgAC8ZTBFCOEGTh6SPKdr-8+jBBTVPNlxanvNqLmwooyxzj7HhNcfJaPylvHO+z+uUJifsJROS9pcSqTZKR+orU+J2lXOtNkZn7v9fdKV7M502IISSPpBZ9FSmZvH04ryDL-HmnGi1KtCmD6d08gn-hDOZ37HYP4AFoAv43jizu7irLA-4ej-7vSAHbo8j37JKP6IHgFKpv4a4f7xpUQSDODf7AFJZgEQHYHdYf5HhI7ejwEQBgFYFQHeYf6ETUG-4IEt5oCJbx7oEcFcwiLxo+gYSRgGBC67zUAA5ZZA5szu7Q4Ob9ZlYVYGCq5Y6ja+a4EvwMD-ZGYR7WY5bSFMHg46GHBQ69Yw5+bw7dJhLI6E7o7SiY4taqENZ44E6KEk5E7kGlpU7+YwH-qKaM5SpH71is5Zoc506KY860AMGfjQFpaEAi7y4EAS7qbGqJa0YJHD5ooIC8IwDuqerhBWHEHcGkHVIwADYZFzYtYzaCAVGI7YgrZrYLYkEYF7ZwF-50EYGnbAjuAkZXYrbOrVg3ysG0H0Gv6MHv79EFirwoBRj5ytHsEfRRH3gxEHpEL2oq4Y41ZNG8GIg2ra6RrRpcAG4JpJo-geGirm42Y8jR4ZpZryDB6-jixh7v6LH1AxGe5Xi54ga+4jrSQcJT5eQ1BtDhBoDoBOgiJtBRg0jUBdTSJ2hTAOhgl2goBQhtDFA-HsIeCfFPpNItI1BYl0a6pfJ8HtF6iwkCG4BCFvBZQ-FfrNr4kJGdruBb6b5sbb6UaEl5JlwsoQCkIQBmIQDXD6KlZUi5YQA+j4CcASAqDrD0A5IilsyZKymxrMiATpA4C7JPg8nABeyGAQCESHBwAaitSy4UB2ATIsLqjRpE4QBRg5KYiml6jrBiC8marao974BJH76wi3DpDX65acmECdTTDhr7H65xrHHG7UAoCEAABC0a1AUIue6AggtuvagZ6wqR-eeez6LSxQoZhx4ZRuNAWIGgkJsI+AmmSZaAZS8gAimZZ62ZOJr6+ZsahuJxi0KA7gVZ+Agg6QfZae5wWEYSNZWIki1S+J+eLS3g6YSqI5dZOetGk5r65aXunCiG1AvY9hKiK5V4ehlEKiEAYAGJUheW5egO66pUsJUeaYyI7AZ5khF5FsYgwSLae57uYgqBMSFx15EAL5f4b5TBEA-+JKWUx5fAAF8eFxj5Mgt5naKYz22yUw+2EWO28e-aIBaBCFXMRhLarx8J4QsAZkbqeQL58FcIxJZFsIwIfZ6Q4FwOMARFiIaYeQn5TilF2FFehwsi+FaAN4jFgMpFWFsJlF1FfZ-5wO14jFzFAo0sWFYgngW2H2qFGuqWR53kkBRAEFGu35VeN5eJRRzRFx3F1AzgBFklxF4gpFpBwlYBol6Q4l0h5lTFH5GFMSPBz+UF-OPFhFFlSgVlhlIi7lSBNFdF0hDFbq0lSg0spBCiHxyYOA3k25x56xdhNWFmOFUONQ6VnFyqmwvC3lSAvlL5yq2ALl2V55XlJlBF4VAWygxViQlFGiyFnW4V6FKVjg9h2RRF7FyIfMUVnCRlUB3lTlAllxGAyhLW5lQV4BOxkVygwFIKoF6lNhShGxWOMZZiVwti95oW0FhwIi159WZi2AuKFxb5G1W1NiTBsFF121kFOFe1z5KOihfAE1vCt1V18eR1J1tVQFEAH1O1leVxc1v5z1aOBgf4b1ANX1vmx1uKv14gC1reJQx5K1r1a1LW0N2lD1uGG631p1OF51m1d1OBsFnEcNFxKY3lCFdl3ljShAkVL5CFFxZcJF-1xNn1Gu7FFx3lvKdi3ko19NlNINdVD60o-NskRsfejgEtHgONbu3lNVRsllEAQthN9F-FPMoNK1b14VwI9NCN8MSNaAytf1fN9hg10Rw1mtPML5atOVb5I1WtKtOtGNvC5l+tjAhthMxtptiNdBJKgMvt81AdIKLlkYi1KNy1ihkNbtWN3WOlwN+NZ1wO8dBWZNFNOF01tNVVvF9NjN7Nl1WxHlOFvNyYstbN9tgO6Yo15tAtMlYtMtFt8teFudPl3tflqtjAKdYVNtUML5rtqVWOetXdRAHdIdEd4QUtjduAstltSx1tvl-do9boDtElfd8MA9ihb1HtK949gQwdRtodU9E9gd-tk9aAsVs0bhBAvux5N9ZO-Oad5xLdAgeNsNJ1LN6Z0whhOVmV7Ag2w271HNjhcNyIAAZAoMBHuU1eTSdW1Z4B7vfRoW8LIs-Yuhlbpb-YDv-QNTheg25dZS+QQ+hmEiQ1YtNcSfjjampRwitQ-ShuQ55ZbqVMnerdIeQ2DhAFfJxKvMdVYqvDYqVlYlcNgAI5zQnZgzyBnZ-VndTXCDnaZXnYwAXczThazaLeQ9zaXW3XXZLQKHbd3VnSLZ3Qw0-SAzDTTdw7w-w4I8I6I+I0wdPXo3LTlShorRvQYyvT3XlkrVrS+WY27uQ8CFY-Qyg4-W7iPfTcCDw3w2I3YyI-ExI+nX7X9RfabdLbPc3TlcZUo3xUvZvd4+w3lk7c7QE+E4wxYxriEwo2DUTg-bIrvdE9Y3E+I-Y0k9dak+fWfTzIfafWHf0yfWkyBVHXQ64eE2g1U5I3-Vgx-QTWvRw1MwVjEzY0k+044-dTM9IzUHA-M9XbZRADRXTSo09Vo9ZTo3ky45XUYzlTXV4w01AcEy07Y0I4kxszgUHTPXPa-eIh4wU6LVXZIW+X44Uw89EU82Ex5hE68VE4wCs60wkw44I50z7cfSbVDJk98zk0NW3aU8vYC6Fo7Z469OU1C5M0XTDZC6To00RZ7UQPCy8+s8i+-vvYM+i0fek2y+HSBUAA

### CODE REVIEW

**Comment by thiemok** [comment]
Due to formalistics immutable nature it should be possible to improve the precision of return types of mutating operations like `put` and `remove` even more so that the typing of access operations can completely ensure access to valid fields.
I'll update the types accordingly in my next iteration

**Comment by thiemok** [comment]
I've added support for mutating the `VALUE_TYPE` of `MapForm` through mutating operations like `remove` and `put`.
(The playground is updated)

Something similar should be possible for `ListForm` effectively changing their usage from a list (`VALUE_TYPE[]`) to a tuple (`[ string, number, SOMETHING]`), but I felt that might not reflect the intended usage very well and might make typing iterative functions like `map` a bit harder.
@bripkens I'll leave it up to you which variation you would prefer here

**Comment by bripkens** [comment]
> Something similar should be possible for ListForm effectively changing

This would indeed not be the originally intended usage. How about we start with consistent types for `ListForm` and, if the need ever arises, allow usages of varying types by index? I would guess that the need for this is very low and that proper support for this would make general typed `ListForm` usage harder for many who don't need it.

**Comment by thiemok** [comment]
> How about we start with consistent types for ListForm...
I agree, this however poses one restriction with potentially common js usage patterns.
Given a `ListForm` containing `MapForm`s without a common base type. Access via `getIn` or `updateIn` will not allow paths pointing at specific fields within the contained `MapForm`s.
But I would say this would generally be an antipattern and can be avoided by for example manually providing a common base type for the items of that `ListForm` or replacing the `ListForm` with a `MapForm` if the behaviour of a tuple is more desired.

Since this is a basic asumption made by the typings for nested accessors, should it be mentioned in the documentation somewhere?

**Comment by thiemok** [comment]
I've added basic test for `MapForm` and `ListForm`, tests for nested forms will follow next.

One thing we might want to discuss is the typing for `MapForm.containsKey()`. It currently only allows to provide keys that actually exist on the forms type. We might want to open that up to `string`. That way this method could be used to implement type narrowing for polymorphic forms.

File: formalistic.d.ts [hunk 1]
```diff
@@ -38,79 +38,77 @@ export interface MapFormItems {
...
-  push(item: Item): ListForm;
-  insert(index: number, item: Item): ListForm;
-  set(index: number, item: Item): ListForm;
-  unshift(item: Item): ListForm;
-  remove(index: number): ListForm;
-  get(index: number): Item | undefined;
-  getIn(path: string[]): Item;
-  updateIn(path: string[], updater: (item: Item) => Item): ListForm;
-  setTouched(touched: boolean, opts?: SetTouchedOptions): ListForm;
+  push(item: VALUE_TYPE[number]): ListForm<VALUE_TYPE>;
```
Comment (bripkens) [inline formalistic.d.ts:94]: **Question:** Why does this encode `number`?

Comment (thiemok) [reply formalistic.d.ts:94]: Do you mean the `Value_Type[number] `?

**Comment by bripkens** [comment]
Thank you very much @thiemok! ❤️ 

The change is now published as `2.0.0` with a [note in the changelog](https://github.com/bripkens/formalistic/blob/main/CHANGELOG.md#200).

**Pull request merged**